prov/sockets/src/sock_progress.c

```diff
@@ -2629,10 +2630,12 @@ static void sock_pe_poll(struct sock_pe *pe)
 	if (pe->waittime && ((fi_gettime_ms() - pe->waittime) < sock_pe_waittime))
 		return;
 
-	if (dlist_empty(&pe->tx_list) && dlist_empty(&pe->rx_list))
+	pthread_mutex_lock(&pe->list_lock);
+	if (dlist_empty(&pe->tx_list) && dlist_empty(&pe->rx_list)) {
+		pthread_mutex_unlock(&pe->list_lock);
 		return;
+	}
 
-	pthread_mutex_lock(&pe->list_lock);
 	if (!dlist_empty(&pe->tx_list)) {
```

Comment:
I don't think that I have an opinion about splitting `sock_pe_poll()` the way you suggest.  My only $0.02 is that the initial lock should still be moved up to include the "empty" checks.